File: sdks/go/discovery_test.go
package beachcomber_test

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"

	beachcomber "github.com/NavistAu/beachcomber/sdks/go"
)

func TestDiscoverSocketPath_BeachcomberSocket(t *testing.T) {
	// BEACHCOMBER_SOCKET takes precedence over everything else.
	t.Setenv("BEACHCOMBER_SOCKET", "/custom/path/comb.sock")
	t.Setenv("XDG_RUNTIME_DIR", "/run/user/1000")
	t.Setenv("TMPDIR", "/should-not-be-used")

	got := beachcomber.DiscoverSocketPath()
	if got != "/custom/path/comb.sock" {
		t.Errorf("expected %q, got %q", "/custom/path/comb.sock", got)
	}
}

func TestDiscoverSocketPath_XDG(t *testing.T) {
	// XDG_RUNTIME_DIR resolves unconditionally (no existence probe), matching
	// where the daemon binds.
	t.Setenv("BEACHCOMBER_SOCKET", "")
	t.Setenv("XDG_RUNTIME_DIR", "/run/user/1000")
	t.Setenv("TMPDIR", "/should-not-be-used")

	got := beachcomber.DiscoverSocketPath()
	want := filepath.Join("/run/user/1000", "beachcomber", "sock")
	if got != want {
		t.Errorf("expected %q, got %q", want, got)
	}
}

func TestDiscoverSocketPath_XDGUnset(t *testing.T) {
	// XDG unset: fall through to /tmp. TMPDIR is NOT consulted.
	t.Setenv("BEACHCOMBER_SOCKET", "")
	t.Setenv("XDG_RUNTIME_DIR", "")
	t.Setenv("TMPDIR", "/should-not-be-used")

	got := beachcomber.DiscoverSocketPath()
	uid := os.Getuid()
	want := fmt.Sprintf("/tmp/beachcomber-%d/sock", uid)
	if got != want {
		t.Errorf("expected %q, got %q", want, got)
	}
}

func TestDiscoverSocketPath_TMPDIRIgnored(t *testing.T) {
	// TMPDIR must never influence resolution.
	t.Setenv("BEACHCOMBER_SOCKET", "")
	t.Setenv("XDG_RUNTIME_DIR", "")
	t.Setenv("TMPDIR", "/var/folders/xyz")

	got := beachcomber.DiscoverSocketPath()
	uid := os.Getuid()
	want := fmt.Sprintf("/tmp/beachcomber-%d/sock", uid)
	if got != want {
		t.Errorf("expected %q, got %q", want, got)
	}
}

func TestDiscoverSocketPath_ContainsUID(t *testing.T) {
	t.Setenv("BEACHCOMBER_SOCKET", "")
	t.Setenv("XDG_RUNTIME_DIR", "")

	got := beachcomber.DiscoverSocketPath()
	uid := fmt.Sprintf("%d", os.Getuid())
	if !strings.Contains(got, uid) {
		t.Errorf("expected path to contain uid %s, got %q", uid, got)
	}
}

func TestDiscoverSocketPath_EndsSock(t *testing.T) {
	t.Setenv("BEACHCOMBER_SOCKET", "")
	t.Setenv("XDG_RUNTIME_DIR", "")

	got := beachcomber.DiscoverSocketPath()
	if !strings.HasSuffix(got, "/sock") {
		t.Errorf("expected path to end with /sock, got %q", got)
	}
}

File: sdks/node/tests/discovery.test.ts
import { describe, it, before, after } from 'node:test';
import assert from 'node:assert/strict';
import { discoverSocketPath, getUid } from '../src/discovery.js';

describe('discoverSocketPath', () => {
  let savedSock: string | undefined;
  let savedXdg: string | undefined;
  let savedTmpdir: string | undefined;

  const restore = (key: string, saved: string | undefined): void => {
    if (saved === undefined) {
      delete process.env[key];
    } else {
      process.env[key] = saved;
    }
  };

  before(() => {
    savedSock = process.env['BEACHCOMBER_SOCKET'];
    savedXdg = process.env['XDG_RUNTIME_DIR'];
    savedTmpdir = process.env['TMPDIR'];
  });

  after(() => {
    restore('BEACHCOMBER_SOCKET', savedSock);
    restore('XDG_RUNTIME_DIR', savedXdg);
    restore('TMPDIR', savedTmpdir);
  });

  it('BEACHCOMBER_SOCKET takes precedence over everything', () => {
    process.env['BEACHCOMBER_SOCKET'] = '/custom/path/comb.sock';
    process.env['XDG_RUNTIME_DIR'] = '/run/user/1000';
    process.env['TMPDIR'] = '/should-not-be-used';

    assert.equal(discoverSocketPath(), '/custom/path/comb.sock');
  });

  it('ignores XDG_RUNTIME_DIR entirely', () => {
    delete process.env['BEACHCOMBER_SOCKET'];
    process.env['XDG_RUNTIME_DIR'] = '/run/user/1000';
    process.env['TMPDIR'] = '/should-not-be-used';

    const uid = getUid();
    assert.equal(discoverSocketPath(), `/tmp/beachcomber-${uid}/sock`);
  });

  it('falls back to /tmp when XDG_RUNTIME_DIR is unset', () => {
    delete process.env['BEACHCOMBER_SOCKET'];
    delete process.env['XDG_RUNTIME_DIR'];
    process.env['TMPDIR'] = '/should-not-be-used';

    const uid = getUid();
    assert.equal(discoverSocketPath(), `/tmp/beachcomber-${uid}/sock`);
  });

  it('ignores TMPDIR entirely', () => {
    delete process.env['BEACHCOMBER_SOCKET'];
    delete process.env['XDG_RUNTIME_DIR'];
    process.env['TMPDIR'] = '/var/folders/xyz';

    const uid = getUid();
    assert.equal(discoverSocketPath(), `/tmp/beachcomber-${uid}/sock`);
  });
});

describe('getUid', () => {
  it('returns a string', () => {
    const uid = getUid();
    assert.equal(typeof uid, 'string');
  });

  it('returns a non-empty string', () => {
    const uid = getUid();
    assert.ok(uid.length > 0);
  });

  it('returns a numeric-looking string', () => {
    const uid = getUid();
    assert.ok(/^\d+$/.test(uid), `expected numeric string, got: ${uid}`);
  });
});
